From 22a7567f0f673d5936dd22a743a88a56e6edfd94 Mon Sep 17 00:00:00 2001
From: Roman Stratiienko <r.stratiienko@gmail.com>
Date: Sun, 9 Apr 2023 23:21:03 +0300
Subject: [PATCH 11/11] codecs: Add and enable the ffmpeg-codec2 project

Use ffmpeg for software codecs. It should improve decoding performance
by utilizing more cores compared to Google's default SW codecs.

Signed-off-by: Roman Stratiienko <r.stratiienko@gmail.com>
---
 common/codecs/board.mk                        | 10 +++++++
 common/codecs/device.mk                       | 27 ++++++++++++++++++-
 ...work_compatibility_matrix_ffmpegcodec2.xml | 11 ++++++++
 common/codecs/media_codecs.xml                |  5 ++++
 common/codecs/sepolicy/vendor/file_contexts   |  2 ++
 5 files changed, 54 insertions(+), 1 deletion(-)
 create mode 100644 common/codecs/device_framework_compatibility_matrix_ffmpegcodec2.xml
 create mode 100644 common/codecs/media_codecs.xml
 create mode 100644 common/codecs/sepolicy/vendor/file_contexts

diff --git a/common/codecs/board.mk b/common/codecs/board.mk
index 9358763..fed5d9b 100644
--- a/common/codecs/board.mk
+++ b/common/codecs/board.mk
@@ -4,8 +4,18 @@
 #
 # Copyright (C) 2022 Roman Stratiienko (r.stratiienko@gmail.com)
 
+BCDC_PATH := $(patsubst $(CURDIR)/%,%,$(dir $(abspath $(lastword $(MAKEFILE_LIST)))))
+
 BOARD_BUILD_AOSPEXT_DAV1D := true
 BOARD_DAV1D_SRC_DIR := glodroid/vendor/dav1d
 
 BOARD_BUILD_AOSPEXT_FFMPEG := true
 BOARD_FFMPEG_SRC_DIR := glodroid/vendor/ffmpeg
+
+BOARD_BUILD_AOSPEXT_FFMPEGCODEC2 := true
+BOARD_FFMPEGCODEC2_SRC_DIR := glodroid/vendor/ffmpeg_codec2
+
+BOARD_VENDOR_SEPOLICY_DIRS += $(BCDC_PATH)/sepolicy/vendor
+
+DEVICE_FRAMEWORK_COMPATIBILITY_MATRIX_FILE += \
+    $(BCDC_PATH)/device_framework_compatibility_matrix_ffmpegcodec2.xml \
diff --git a/common/codecs/device.mk b/common/codecs/device.mk
index e04adfa..b60c263 100644
--- a/common/codecs/device.mk
+++ b/common/codecs/device.mk
@@ -7,13 +7,21 @@
 PRODUCT_PACKAGES += \
     android.hardware.media.c2@1.0-service \
 
+PRODUCT_PACKAGES += \
+    android.hardware.media.c2@1.2-service-ffmpeg \
+    android.hardware.media.c2@1.2-service-ffmpeg.rc \
+    android.hardware.media.c2@1.2-service-ffmpeg.xml \
+    media_codecs_ffmpeg_c2.xml \
+
 # Create input surface on the framework side
 PRODUCT_VENDOR_PROPERTIES += \
     debug.stagefright.c2inputsurface=-1 \
 
+GD_MEDIACODECS_FILE ?= $(LOCAL_PATH)/media_codecs.xml
+
 # Copy media codecs config file
 PRODUCT_COPY_FILES += \
-    frameworks/av/media/libstagefright/data/media_codecs_google_c2.xml:$(TARGET_COPY_OUT_VENDOR)/etc/media_codecs.xml \
+    $(GD_MEDIACODECS_FILE):$(TARGET_COPY_OUT_VENDOR)/etc/media_codecs.xml \
     frameworks/av/media/libstagefright/data/media_codecs_google_c2_video.xml:$(TARGET_COPY_OUT_VENDOR)/etc/media_codecs_google_c2_video.xml \
     frameworks/av/media/libstagefright/data/media_codecs_google_c2_audio.xml:$(TARGET_COPY_OUT_VENDOR)/etc/media_codecs_google_c2_audio.xml \
     $(LOCAL_PATH)/media_profiles_V1_0.xml:$(TARGET_COPY_OUT_VENDOR)/etc/media_profiles_V1_0.xml \
@@ -22,3 +30,20 @@ PRODUCT_COPY_FILES += \
 PRODUCT_COPY_FILES += \
     $(LOCAL_PATH)/mediaswcodec.policy:$(TARGET_COPY_OUT_VENDOR)/etc/seccomp_policy/mediaswcodec.policy \
     $(LOCAL_PATH)/mediacodec.policy:$(TARGET_COPY_OUT_VENDOR)/etc/seccomp_policy/mediacodec.policy \
+
+# Codec2.0 poolMask:
+#   ION(16)
+#   GRALLOC(17)
+#   BUFFERQUEUE(18)
+#   BLOB(19)
+#   V4L2_BUFFERQUEUE(20)
+#   V4L2_BUFFERPOOL(21)
+#   SECURE_LINEAR(22)
+#   SECURE_GRAPHIC(23)
+#
+# For linear buffer allocation:
+#   If ION is chosen, then the mask should be 0xf50000
+#   If GRALLOC is chosen, then the mask should be 0xf60000
+#   If BLOB is chosen, then the mask should be 0xfc0000
+PRODUCT_PROPERTY_OVERRIDES += \
+    debug.stagefright.c2-poolmask=0x0c0000 \
diff --git a/common/codecs/device_framework_compatibility_matrix_ffmpegcodec2.xml b/common/codecs/device_framework_compatibility_matrix_ffmpegcodec2.xml
new file mode 100644
index 0000000..6c335a7
--- /dev/null
+++ b/common/codecs/device_framework_compatibility_matrix_ffmpegcodec2.xml
@@ -0,0 +1,11 @@
+<compatibility-matrix version="1.0" type="framework">
+    <hal format="hidl">
+        <name>android.hardware.media.c2</name>
+        <transport>hwbinder</transport>
+        <version>1.2</version>
+        <interface>
+            <name>IComponentStore</name>
+            <instance>ffmpeg</instance>
+        </interface>
+    </hal>
+</compatibility-matrix>
diff --git a/common/codecs/media_codecs.xml b/common/codecs/media_codecs.xml
new file mode 100644
index 0000000..d2ff08f
--- /dev/null
+++ b/common/codecs/media_codecs.xml
@@ -0,0 +1,5 @@
+<MediaCodecs>
+    <Include href="media_codecs_ffmpeg_c2.xml" />
+    <Include href="media_codecs_google_c2_audio.xml" />
+    <Include href="media_codecs_google_c2_video.xml" />
+</MediaCodecs>
\ No newline at end of file
diff --git a/common/codecs/sepolicy/vendor/file_contexts b/common/codecs/sepolicy/vendor/file_contexts
new file mode 100644
index 0000000..5b7656d
--- /dev/null
+++ b/common/codecs/sepolicy/vendor/file_contexts
@@ -0,0 +1,2 @@
+# Ffmpeg codec2
+/vendor/bin/hw/android\.hardware\.media\.c2@1\.2-service-ffmpeg(.*)?		u:object_r:mediacodec_exec:s0
-- 
2.37.2

